### PR TITLE
Support for CSR in PEM added

### PR DESCRIPTION
**Pull request by encedo**
Hi Titouan,

here is my pull request with a small patch that supports getting a certificate based on CSR in PEM format (in my case generated inside HSM).

Review and merge it if thinks it's useful. In my case it is :) HSMs (many of them, small USB tokens) sends CSR in PEM and backend worker based on CSR and the last issued certificate renew a certificate by selecting one of the three supported providers: BuyPASS, Let's Encrypt, and now ZeroSSL.

Best
Chris

P.S.
Another small enhancement should be added to acmephp - a DNS TXT record generation. We do this:

```
$challenges = $order->getAuthorizationChallenges($cn);
...
$hash = hash("sha256", $challenge->getPayload(), true);
$txt = rtrim( strtr(base64_encode( $hash), '+/', '-_'), '=');
```

and now $txt holds a value required by `_acme-challenge.$cn` DNS TXT record. Our system is fully automated, DNS is managed by API (we use and recommend DigitalOcean). Think about adding a function like this:

`$txt = $challenge->getDNSValue();`

or similar.

### CODE REVIEW

**Comment by encedo** [comment]
FYI 

to generate CSR in one-line try this (as classic CSR generation procedure do not add SAN which is required by ACME):

`openssl req -new -newkey rsa:2048 -nodes -keyout server.key -out server.csr  -addext "subjectAltName=DNS:example.com" -subj "/CN=example.com"`

**Comment by encedo** [comment]
Just more two updates to make the lib RFC8555 compiant.

File: src/Ssl/CertificateRequest.php [hunk 1]
```diff
@@ -24,19 +24,37 @@ class CertificateRequest
     /** @var KeyPair */
     private $keyPair;
 
-    public function __construct(DistinguishedName $distinguishedName, KeyPair $keyPair)
+    /** @var CsrPem */
+    private $csrPem;
+
+    public function __construct(DistinguishedName $distinguishedName, KeyPair $keyPair, string $csr_pem = null)
```
Comment (tgalopin) [inline src/Ssl/CertificateRequest.php:30]: Could you use a camelCase naming?

File: src/Ssl/CertificateRequest.php [hunk 2]
```diff
@@ -24,19 +24,37 @@ class CertificateRequest
...
         $this->distinguishedName = $distinguishedName;
         $this->keyPair = $keyPair;
+        $this->csrPem = $csr_pem;
     }
 
-    public function getDistinguishedName(): DistinguishedName
+    /**
+     * @return DistinguishedName
+     */
+    public function getDistinguishedName()
```
Comment (tgalopin) [inline src/Ssl/CertificateRequest.php:40]: You should revert this change

File: src/Ssl/CertificateRequest.php [hunk 3]
```diff
@@ -24,19 +24,37 @@ class CertificateRequest
...
     {
         return $this->distinguishedName;
     }
 
-    public function getKeyPair(): KeyPair
+    /**
+     * @return KeyPair
+     */
+    public function getKeyPair()
```
Comment (tgalopin) [inline src/Ssl/CertificateRequest.php:48]: You should revert this change

File: src/Ssl/CertificateRequest.php [hunk 4]
```diff
@@ -24,19 +24,37 @@ class CertificateRequest
...
     {
         return $this->keyPair;
     }
+
+    /**
+     * @return CsrPem
+     */
+    public function getCsrPem()
```
Comment (tgalopin) [inline src/Ssl/CertificateRequest.php:56]: You can use return typehints as you're targeting 2.0 which supports PHP 7.2+

File: src/Ssl/CertificateRequest.php [hunk 5]
```diff
@@ -24,19 +24,37 @@ class CertificateRequest
     /** @var KeyPair */
     private $keyPair;
 
-    public function __construct(DistinguishedName $distinguishedName, KeyPair $keyPair)
+    /** @var CsrPem */
```
Comment (tgalopin) [inline src/Ssl/CertificateRequest.php:27]: Isn't it a string?

File: src/Core/AcmeClient.php [hunk 6]
```diff
@@ -328,6 +332,7 @@ protected function requestResource(string $method, string $resource, array $payl
     private function createCertificateResponse(CertificateRequest $csr, string $certificate): CertificateResponse
     {
         $certificatesChain = null;
+        $certificate = str_replace("-----END CERTIFICATE-----\n-----BEGIN CERTIFICATE-----", "-----END CERTIFICATE-----\n\n-----BEGIN CERTIFICATE-----", $certificate);
```
Comment (tgalopin) [inline src/Core/AcmeClient.php:335]: What's the aim of that?

File: src/Core/AcmeClient.php [hunk 7]
```diff
@@ -155,7 +155,11 @@ public function finalizeOrder(CertificateOrder $order, CertificateRequest $csr,
         if (\in_array($response['status'], ['pending', 'processing', 'ready'])) {
             $humanText = ['-----BEGIN CERTIFICATE REQUEST-----', '-----END CERTIFICATE REQUEST-----'];
 
-            $csrContent = $this->csrSigner->signCertificateRequest($csr);
+            $csrContent = $csr->getCsrPem();
```
Comment (tgalopin) [inline src/Core/AcmeClient.php:158]: You can inline this:

```
$csrContent = $csr->getCsrPem() ?: $this->csrSigner->signCertificateRequest($csr);
```

**Review by tgalopin — COMMENTED** [review]
This is a nice PR, thanks! A few comments but I'm not opposed to merging it at all :)

File: src/Ssl/CertificateRequest.php (hunk 1 above)
Comment (encedo) [reply src/Ssl/CertificateRequest.php]: Fixed, my bad :/

File: src/Ssl/CertificateRequest.php (hunk 2 above)
Comment (encedo) [reply src/Ssl/CertificateRequest.php:40]: I should used git not composer first place. All changes (as are mostly one liners) are made using GitHub web. I'm not a pro PHP programmer, just a hacker ;)
Sorry.

File: src/Ssl/CertificateRequest.php (hunk 5 above)
Comment (encedo) [reply src/Ssl/CertificateRequest.php:27]: csrPem variable? Yes, csrPem is a string (base64 encoded PKCS#10).

File: src/Core/AcmeClient.php (hunk 7 above)
Comment (encedo) [reply src/Core/AcmeClient.php:158]: Honestly, I'm old-scholl C programmer (Linux kernel style), don't like inlines like this.

File: src/Core/AcmeClient.php (hunk 6 above)
Comment (encedo) [reply src/Core/AcmeClient.php:335]: True, a little more explanation is needed - the aim is to make `getCertificate()->getPem()` & `getCertificate()->getIssuerCertificate()->getPEM()` working as expected.

Let's Encrypt returns the chain certs separated by a new line. That is why your code `explode("\n\n", $certificate)` works. But regards RFC there is no such a separation, casing the above function to failed when ZeroSSL or BuyPASS CAs are used.

One single (but yes, dirty) line of code makes the functions mentioned working as expected for all three CAs. Otherwise more sophisticated parser is needed to split leave certificate from all the rest.

**Comment by encedo** [comment]
Hi Titouan,

all my changes are done as a result of the research of the ACME lib to work with other CAs . As I said I'm not a pro PHP programmer. I just wanted to share my findings ;)

Small suggestion - consider adding BuyPass CA test environment (test endpoint, more [here](https://community.buypass.com/t/63d4ay/buypass-go-ssl-endpoints)) to the test scripts. ZeroSSL is not running a stage endpoint (they are Sectigo resellers, that's why the EAB was needed ;)

Chris

File: src/Ssl/CertificateRequest.php (hunk 5 above)
Comment (tgalopin) [reply src/Ssl/CertificateRequest.php:27]: Then the php doc is wrong :)

Comment (encedo) [reply src/Ssl/CertificateRequest.php:27]: Not sure if we are talking about the same thing :/ php doc?

**Comment by tgalopin** [comment]
If that's OK for you, I'll finish the pr then :)

**Comment by encedo** [comment]
Finish/merge and bump version to 2.0.1 :)

**Comment by tgalopin** [comment]
Replaced by https://github.com/acmephp/acmephp/pull/225

Thanks!
